plugin/evm/limitorders/build_block_pipeline.go

```diff
@@ -22,24 +22,55 @@ func NewBuildBlockPipeline(db LimitOrderDatabase, lotp LimitOrderTxProcessor) *B
 func (pipeline *BuildBlockPipeline) Run(lastBlockTime uint64) {
 	pipeline.lotp.PurgeLocalTx()
 	if isFundingPaymentTime(lastBlockTime, pipeline.db) {
-		log.Info("BuildBlockPipeline - isFundingPaymentTime")
+		log.Info("BuildBlockPipeline:isFundingPaymentTime")
 		// just execute the funding payment and skip running the matching engine
 		err := executeFundingPayment(pipeline.lotp)
 		if err != nil {
 			log.Error("Funding payment job failed", "err", err)
 		}
 	} else {
-		for _, market := range GetActiveMarkets() {
-			pipeline.runLiquidationsAndMatchingForMarket(market)
+		// fetch the underlying price and run the matching engine
+		underlyingPrices, err := pipeline.lotp.GetUnderlyingPrice()
+		if err != nil {
+			log.Error("could not fetch underlying price", "err", err)
+		} else {
+			for i, market := range GetActiveMarkets() {
+				pipeline.runLiquidationsAndMatchingForMarket(market, underlyingPrices[i])
```

Comment:
we are assuming first element of GetActiveMarket will be same first element of amm.we are making a implicit connection here with amm ids for market  from contract response with GetActiveMarkets. 

We can return map of ammid => underlyingPrice in contract and then use it to iterate here. We use ammid for storing new orders so it should work internally. 
Also we may populate GetActiveMarkets as well from contracts but it can be done separately

Reply:
what you say is correct, but we are ok to go with this assumption for now - that contracts will return the prices in the same order. decoding a mapping in go will be hard. already took me 1 day to decode an array